File: Depth-First Search/Binary tree/Find paths/LET_no563_Binary Tree Tilt.py
# The same as LintCode no1172_Binary Tree Tilt

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findTilt(self, root: TreeNode) -> int:
        res, _ = self.dfs(root)
        return res
    
    def dfs(self, root):
        if not root:
            return 0, 0
        left_diff, left_total = self.dfs(root.left)
        right_diff, right_total = self.dfs(root.right)
        return left_diff + right_diff + abs(left_total - right_total), left_total + right_total + root.val

File: Depth-First Search/Binary tree/Find paths/Easy/LET_no563_Binary Tree Tilt.cpp
// The same as LintCode no1172_Binary Tree Tilt

/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */

class Solution {
public:
    /**
     * @param root: the root
     * @return: the tilt of the whole tree
     */
    int res = 0;
    int findTilt(TreeNode * root) {
        // Write your code here
        dfs(root);
        return res;
    }
    int dfs(TreeNode* root) {
        if (!root)
            return 0;
        
        int leftSum = dfs(root->left);
        int rightSum = dfs(root->right);
        res += abs(leftSum - rightSum);

        return leftSum + rightSum + root->val;
    }
};